File: src/401_500/0416_partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution(object):
	# 递归+剪枝
	def canPartition(self, nums):
		"""
		:type nums: List[int]
		:rtype: bool
		"""				
		if not nums:
			return False
		total = sum(nums)
		if total%2==1:
			return False
		total /= 2
		n = len(nums)
		nums = sorted(nums,reverse=True)
		def dfs(index, remain):
			if remain==0:
				return True
			if index<n and remain<nums[index]:
				return False
			for i in xrange(index,n):
				if dfs(i+1, remain-nums[i]):
					return True
			return False
		return dfs(0,total)
		
	
	# 动态规划实现
	def canPartition(self, nums):
		"""
		:type nums: List[int]
		:rtype: bool
		"""				
		if not nums:
			return False
		total = 0 
		for i in nums:
			total += i
		if total%2==1:
			return False
		total /= 2
		dp = [False]*(total+1)
		dp[0] = True
		for c in nums:
			for i in xrange(total,c-1,-1):
				if dp[i-c]:
					dp[i] = True
		return dp[-1]

File: src/401_500/0416_partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
	public boolean canPartition(int[] nums) {
		if(nums==null){
			return false;
		}
		int total = 0;
		for(int i : nums) {
			total += i;
		}
		if(total%2==1) {
			return false;
		}
		total /= 2;
		int[][] mem = new int[nums.length][total+1];
		for(int i=0;i<nums.length;i++) {
			java.util.Arrays.fill(mem[i],-1);
		}
		return dfs(mem,nums,0,total);
	}
	
	boolean dfs(int[][] mem, int[] nums, int index, int sum) {
		if(sum==0) {
			return true;
		}
		if(sum<0 || index==nums.length) {
			return false;
		}
		if(mem[index][sum]!=-1) {
			return mem[index][sum]==1;
		}
		boolean a = dfs(mem, nums, index+1, sum-nums[index]);
		boolean b = dfs(mem, nums, index+1, sum);
		mem[index][sum] = (a||b)? 1 : 0;
		return mem[index][sum]==1;
	}
}		